nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

```diff
@@ -107,6 +112,74 @@ def write_entry(profile, generation, machine_id):
+        efi_file_relative = "EFI/nixos/nixos-%s-generation-%d.efi" % (profile, generation)
+        efi_file = "@efiSysMountPoint@/%s" % (efi_file_relative)
+    else:
+        entry_file = "@efiSysMountPoint@/loader/entries/nixos-generation-%d.conf" % (generation)
+        efi_file_relative = "EFI/nixos/nixos-generation-%d.efi" % (generation)
+        efi_file = "@efiSysMountPoint@/%s" % (efi_file_relative)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        append_initrd_secrets = profile_path(profile, generation, "append-initrd-secrets")
+        if os.path.exists(append_initrd_secrets):
+            initrd = f"{tmpdir}/initrd"
+            shutil.copyfile(
+                profile_path(profile, generation, "initrd"),
+                initrd
+            )
+
+            subprocess.check_call([append_initrd_secrets, "@efiSysMountPoint@%s" % (initrd)])
+        else:
+            initrd = profile_path(profile, generation, "initrd"),
+
+        generation_dir = os.readlink(system_dir(profile, generation))
+        tmp_path = "%s.tmp" % (efi_file)
+
+        kernel_params = "systemConfig=%s init=%s/init " % (generation_dir, generation_dir)
+        with open("%s/kernel-params" % (generation_dir)) as params_file:
+            kernel_params = kernel_params + params_file.read()
+        kernel_param_file = "%s.kernel_params.tmp" % (efi_file)
+        with open(kernel_param_file, 'w') as f:
+            f.write(kernel_params)
+
+        subprocess.check_call([
+            "@binutils@/bin/objcopy",
+            "--add-section", ".osrel={}/etc/os-release".format(generation_dir), "--change-section-vma", ".osrel=0x20000",
+            "--add-section", ".cmdline={}".format(kernel_param_file),           "--change-section-vma", ".cmdline=0x30000",
+            "--add-section", ".linux={}/kernel".format(generation_dir),         "--change-section-vma", ".linux=0x40000",
+            "--add-section", ".initrd={}".format(initrd),                       "--change-section-vma", ".initrd=0x3000000",
+            "{}/sw/lib/systemd/boot/efi/linuxx64.efi.stub".format(generation_dir),
+            tmp_path
+        ])
+        sign_path(tmp_path, efi_file)
+
+        entry_tmp = entry_file + ".tmp";
+        with open(entry_tmp, 'w') as fp:
+            fp.write(SECURE_BOOT_ENTRY.format(
+                profile=" [" + profile + "]" if profile else "",
+                generation=generation,
+                efi=efi_file_relative,
+                description=describe_generation(generation_dir)
+            ))
+            if machine_id is not None:
+                fp.write("machine-id %s\n" % machine_id)
+
+        os.rename(entry_tmp, entry_file)
+        os.unlink(tmp_path)
+        os.unlink(kernel_param_file)
+
+def sign_path(src, output):
+    subprocess.check_call([
+        "@sbsigntool@/bin/sbsign",
+        "--key", "@signingKey@",
+        "--cert", "@signingCertificate@",
+        "--output", output,
```

Comment:
I rolled my own version of this using the same sources and ran into a problem that if the disk fills up while sbsign is writing it's output, it will return 0 (success) regardless. This is a big problem, because:
 
a) Since each efi now contains its own separate initrd and by default old versions aren't cleaned up, the boot disk can fill up quite quickly, and
b) The code signs *all* versions. So as soon as the disk does fill up, all generations fail to be signed and are mangled silently by sbsign.

This causes the boot process to have appeared to work, but result in system where several, possibly all generations are unbootable. 

I'd recommend adding the following after the sbsign command to verify the binary was signed correctly before moving on:

    subprocess.check_call([
        "@sbsigntool@/bin/sbverify",
        "--cert", "@signingCertificate@",
        output
    ])

This way the process will error out, and the user will at least get some sort of indication that something went wrong.

Reply:
Good grief, @redfish64! Thank you!

Reply:
I have extended sign_path to sign to a tmpdir, move, verify, move.